### ISSUE TITLE
Optional Chaining has some bugs with array component. (on VSCode)

### ISSUE THREAD

**Issue by bychung** [microsoft/TypeScript#35830]
**TypeScript Version:**  3.7.3 (release ver)

**Search Terms:**

**Code**

```ts
interface ContactList {
  group: string,
  list?: {
    name: string,
    tel: string
  }[]
}

const myList: ContactList = {
  group: 'Friends',
  list: [{
    name: 'James',
    tel: '1234-5678'
  }]
}

// OK code
if (myList.list) {
  myList.list.push({name: 'Amy', tel: '2345-6789'}); // Ok.
  console.log(myList.list[0].name); // Ok.
}

// ERROR code
myList.list.push({name: 'Amy', tel: '2345-6789'}); // cause an error : Object is possibly 'undefined'
console.log(myList.list[0].name); // cause an error: Object is possibly 'undefined' (same as above)
console.log(myList.list?[0].name); // cause an error: Property 'name' does not exist on type 'number[]'
```

**Expected behavior:**
'OK code'

**Actual behavior:**
'Error code'
Didn't use optional chaining on first line of 'ERROR code', but myList has defined list so it is not error. Object is not possible 'undefined', but 'defined'.
Third line of 'ERROR code' shows myList.list is array of {name: string, tel: string} but TypeScript recognize it array of number.

**Playground Link:** 

**Related Issues:**

**Comment by jcalz** [comment]
You're not using optional chaining.  The syntax `myList.list?[0].name` is interpreted as the conditional operator, and `[0]` is a `number[]`.  Also note that you get the error `: expected` after `name`

The optional chaining operator is alwyas `?.`, even before bracketed index access.   Try `myList.list?.[0].name` instead.  (See the [tc39/proposal-optional-chaining FAQ](https://github.com/tc39/proposal-optional-chaining#faq).)

**Comment by bychung** [comment]
> You're not using optional chaining. The syntax `myList.list?[0].name` is interpreted as the conditional operator, and `[0]` is a `number[]`. Also note that you get the error `: expected` after `name`
> 
> The optional chaining operator is alwyas `?.`, even before bracketed index access. Try `myList.list?.[0].name` instead. (See the [tc39/proposal-optional-chaining FAQ](https://github.com/tc39/proposal-optional-chaining#faq).)

Thank you jcalz, myList.list?[0].name is my fault.
How about the first line of my Error code? I don't understand still yet.

**Comment by jcalz** [comment]
It's working as designed.  `myList` is annotated as `ContactList`.  This is not a union type, so there is no [control-flow analysis narrowing, #8010](https://github.com/microsoft/TypeScript/pull/8010) upon initial assignment.  Thus `myList` is just seen as a `ContactList`, and its `list` property is possibly `undefined`.  

Note that this is not a forum for getting help with bugs in your code; it's for reporting bugs in the compiler or making feature suggestions.  If you're doing either of those, you should fill out all the fields in the issue creation template.  "Search Terms", "Playground Link", and "Related Issues" have not been filled out above.  Please either do so or close this issue and consider asking questions elsewhere.  Good luck!

**Comment by bychung** [comment]
> It's working as designed. `myList` is annotated as `ContactList`. This is not a union type, so there is no [control-flow analysis narrowing, #8010](https://github.com/microsoft/TypeScript/pull/8010) upon initial assignment. Thus `myList` is just seen as a `ContactList`, and its `list` property is possibly `undefined`.
> 
> Note that this is not a forum for getting help with bugs in your code; it's for reporting bugs in the compiler or making feature suggestions. If you're doing either of those, you should fill out all the fields in the issue creation template. "Search Terms", "Playground Link", and "Related Issues" have not been filled out above. Please either do so or close this issue and consider asking questions elsewhere. Good luck!

Thank you and sorry for my mistake.
I thought it wasn't an error for previous TypeScript version. Because I coded almost over 100k line, but it's the first time to get challenged like this error.
I'll bear your opinion in mind.

**Comment by bychung** [comment]
It was closed, but in my experience it is something weird.. so I tried code like below,
```
interface ContactList {
  group?: string,
  list?: {
    name: string,
    tel: string
  }[]
}

const myList: ContactList = {
  group: 'Friends',
}

myList.list = [{
  name: 'James',
  tel: '1234-5678'
}]
```

and call this line, it works well.
```
myList.list.push({name: 'Amy', tel: '2345-6789'}); // no error.
```

I think it is same that initializing and assigning, TS compiler recognize only assigning case to be defined. Is it intended behavior? Maybe so, compiler need to be smarter than now I think.

**Comment by jcalz** [comment]
It is still working as designed.  `myList` is of a non-union type, but `myList.list` is a union-typed property.  Assigning to `myList` therefore does not do control-flow narrowing, but assigning to `myList.list` does.  It would be great if the compiler could do control-flow narrowing even in non-union types, but it might not be worth degrading compiler performance to do so.

This is possibly a duplicate of or related to #8513, especially [this comment](https://github.com/microsoft/TypeScript/issues/8513#issuecomment-217871478).

**Comment by bychung** [comment]
Got the control-flow narrowing, but still there's a point.
You said control-flow narrowing is only applied to union type except non-union type, and my point is why the compiler does not investigate inner non-union(`myList`) that has union property(`myList.list`). Maybe I'm wrong, but I know the compiler has already got the structure of non-union(`myList`) and there's a union property(`myList.list`).

If the performance is the problem, I can't understand why is so expensive because it costs only one check for `myList.list` in case the assigning at declaration. (compiler do not additionally investigate which property in `myList` is union, because already knows it). If that costs so high, this example(https://github.com/microsoft/TypeScript/issues/35830#issuecomment-568372460) also costs high.

And programmers maybe do like this(https://github.com/microsoft/TypeScript/issues/35830#issuecomment-568372460), then performance is same. 

In conclusion, is the compiler costs really high to do the control-flow narrowing only the case for declaration of non-union type variables only that has union property?

PS. Sorry for my English. Not a native man. T.T

> It is still working as designed. `myList` is of a non-union type, but `myList.list` is a union-typed property. Assigning to `myList` therefore does not do control-flow narrowing, but assigning to `myList.list` does. It would be great if the compiler could do control-flow narrowing even in non-union types, but it might not be worth degrading compiler performance to do so.
> 
> This is possibly a duplicate of or related to #8513, especially [this comment](https://github.com/microsoft/TypeScript/issues/8513#issuecomment-217871478).

**Comment by jcalz** [comment]
At this point you're talking to only me, because the issue is closed.  If you want to reopen it, you should edit the initial post and complete the template.  Good luck.
